* DSC550
* Week 3
* Dan Morrow

### Week 3 assignment 

Part 1: Using the TextBlob Sentiment Analyzer
* Import the movie review data as a data frame and ensure that the data is loaded properly.
* How many of each positive and negative reviews are there?
* Use TextBlob to classify each movie review as positive or negative. Assume that a polarity score greater than or equal to zero is a positive sentiment and less than 0 is a negative sentiment.
* Check the accuracy of this model. Is this model better than random guessing?
* For up to five points extra credit, use another prebuilt text sentiment analyzer, e.g., VADER, and repeat steps (3) and (4).

______

Part 2: Prepping Text for a Custom Model
If you want to run your own model to classify text, it needs to be in proper form to do so. The following steps will outline a procedure to do this on the movie reviews text.


* Convert all text to lowercase letters.
* Remove punctuation and special characters from the text.
* Remove stop words.
* Apply NLTK’s PorterStemmer.
* Create a bag-of-words matrix from your stemmed text (output from (4)), where each row is a word-count vector for a single movie review (see sections 5.3 & 6.8 in the Machine Learning with Python Cookbook).
* Display the dimensions of your bag-of-words matrix. The number of rows in this matrix should be the same as the number of rows in your original data frame.
* Create a term frequency-inverse document frequency (tf-idf) matrix from your stemmed text, for your movie reviews (see section 6.9 in the Machine Learning with Python Cookbook).
* Display the dimensions of your tf-idf matrix. These dimensions should be the same as your bag-of-words matrix.


Additional Comments



The bag-of-words and tf-idf matrices are stored as sparse matrices because most entries are zero.
Each row in the bag-of-words/tf-idf matrices corresponds to a movie review.
The columns in the bag-of-words/tf-idf matrices correspond to unique words appearing in the movie reviews.
Entries in the bag-of-words matrix are the number of times a word appears in a review.
Entries in the tf-idf matrix are numbers representing the word importance in a review.
The bag-of-words/tf-idf matrices are possible feature (input) matrices for model building.
We will revisit this preprocessed text data to build a custom model in the future.
____________

#### Exercise 1. Load the movie Review Dataset

In this step, I will load the labeled movie review dataset into a pandas DataFrame, to then be inspected. 
The output will show the first 5 lines of data returned. 

Dataset:
The dataset contains three columns: `id`, `sentiment`, and `review`. 
The `sentiment` column represents whether a movie review is positive or negative, where a value of 1 indicates a positive review and a value of 0 indicates a negative review. 
The `review` column contains the full text of each movie review.

In [1]:
import pandas as pd

#Load dataset
df = pd.read_csv("labeledTrainData.tsv", sep="\t")

#Output first 5 rows
df.head()

,id,sentiment,review
0,5814_8,1,With all this stuff going down at the moment w...
1,2381_9,1,"\The Classic War of the Worlds\"" by Timothy Hi..."
2,7759_3,0,The film starts with a manager (Nicholas Bell)...
3,3630_4,0,It must be assumed that those who praised this...
4,9495_8,1,Superbly trashy and wondrously unpretentious 8...


#### Exercise 1.a. How many of each positive and negative reviews are there?

In this step, I will count how many positive and negative reviews are present in the dataset. 

The output will show the numbers of postive and negative reviews, with 1 indicating positive reviews, and 0 as negative reviews. 

The ouput 

In [2]:
#Count each sentiment value
sentiment_counts = df["sentiment"].value_counts()

#Output Sentiment counts
sentiment_counts

sentiment
1    12500
0    12500
Name: count, dtype: int64

#### Exercise 1.a explanation: 
The dataset contains an equal number of positive and negative reviews, with 12,500 of each. This shows a balanced dataset, which indicates there isn't a biased toward one class. 

#### Exercise 1.b Use TextBlob to classify each movie review as positive or negative. Assume that a polarity score greater than or equal to zero is a positive sentiment and less than 0 is a negative sentiment.

In this step, I will use the TextBlob library to analyze the sentiment of each movie review. 
TextBlob provides a polarity score ranging from -1 to 1, where negative values indicate negative sentiment and positive values indicate positive sentiment.



In [17]:
from textblob import TextBlob

#Create function to classify sentiment with TextBlob
def get_sentiment(review):
    
    #Get polarity score
    polarity = TextBlob(review).sentiment.polarity
    
    #Apply classification rule
    if polarity >= 0:
        return 1
    else:
        return 0

#Apply function to all reviews
df["predicted_sentiment"] = df["review"].apply(get_sentiment)

#Output first 5 rows 
df[["sentiment", "predicted_sentiment"]].head()

,sentiment,predicted_sentiment
0,1,1
1,1,1
2,0,0
3,0,1
4,1,0


#### Exercise 1.b explanation:

The TextBlob sentiment analyzer was applied to each review, and then the predicted sentiment was stored in a new column called `predicted_sentiment`. 
Comparing the actual and predicted values in the sample output shows that TextBlob correctly identifies some reviews and misclassifies others. 

In 1.c we will calculate overall accuracy to determine how well the model performs.

#### Exercise 1.C Check the accuracy of this model. Is this model better than random guessing?

In this step, I will compare the actual sentiment values to the predicted sentiment values from TextBlob. This will allow me to calculate the model's accuracy and determine whether it performs better than random guessing.

Since the dataset is balanced, random guessing should produce an accuracy of about 50% of the time. 

The output will show us the accuracy value, which is the proportion of reviews that TextBlob classified correctly. Because the dataset contains an equal number of positive and negative reviews, a model would need to perform better than 0.50 accuracy to be considered better than random guessing. If the accuracy is greater than 0.50, then TextBlob performs better than random guessing on this dataset.

In [5]:
#Calculate proportion of correct predictions
accuracy = (df["sentiment"] == df["predicted_sentiment"]).mean()

#Output accuracy
accuracy

np.float64(0.68524)

#### Exercise 1.c explanation:
The TextBlob model achieved an accuracy of approximately 0.685, meaning it correctly classified about 68.5% of the movie reviews. Since the dataset is evenly balanced between positive and negative reviews, random guessing would result in an accuracy of around 50%. 

Because the model performs significantly better than 0.50, it tells that that TextBlob is a meaningful improvement over random guessing. Although this is a relatively strong accuracy, it also shows that TextBlob is not perfect and misclassifies a portion of the reviews.

_______________

#### Extra Credit 1: use another prebuilt text sentiment analyzer, e.g., VADER, and repeat steps (3) and (4).

Similar to the first exercise using TextBlob, I will now use the VADER (Valence Aware Dictionary and sEntiment Reasoner) sentiment analyzer to classify each movie review. 

The compound score from VADER ranges from -1 to 1. For this assignment, I will classify reviews as positive if the compound score is greater than or equal to 0, and negative if it is less than 0.


#### Extra Credit 1 Use Vader to classify each movie review as positive or negative. Assume that a polarity score greater than or equal to zero is a positive sentiment and less than 0 is a negative sentiment.

In this step, I will use the Vader library to analyze the sentiment of each movie review. Vader, similar to TextBlob, provides a polarity score ranging from -1 to 1, where negative values indicate negative sentiment and positive values indicate positive sentiment.

In [18]:
from nltk.sentiment import SentimentIntensityAnalyzer

#Initialize VADER
sia = SentimentIntensityAnalyzer()

#Function to classify sentiment
def get_vader_sentiment(review):
    score = sia.polarity_scores(review)
    
    if score['compound'] >= 0:
        return 1
    else:
        return 0

#Apply function
df["vader_sentiment"] = df["review"].apply(get_vader_sentiment)

#Ouput first 5 rows
df[["sentiment", "vader_sentiment"]].head()

,sentiment,vader_sentiment
0,1,0
1,1,1
2,0,0
3,0,0
4,1,1


In [9]:
vader_accuracy = (df["sentiment"] == df["vader_sentiment"]).mean()
vader_accuracy

np.float64(0.69356)

#### Extra Credit 1 explanation:
The VADER sentiment analyzer achieved an accuracy of approximately 0.694,  which is slightly higher than the TextBlob accuracy of 0.685. Both models performed better than random guessing (0.50), indicating that they are both effective at capturing sentiment in the movie reviews.

While the difference in accuracy is relatively small, VADER appears to perform slightly better on this dataset. This may be because VADER is specifically designed for sentiment analysis and is optimized for interpreting text-based data (according to the reading and information online). Overall, both models provide useful results, but VADER demonstrates a slight advantage in this case.

______________

#### Part 2 Prepping text for a custom model
This is using the same dataset from the above exercise. 

#### Exercise 2.a Convert all text to lowercase letters
This is done to ensure all text is treated the same, the information will be placed in a new column titled "review_lower".

In [10]:
#Convert review text to lowercase and store in new column
df["review_lower"] = df["review"].str.lower()

#Output first few rows
df["review_lower"].head()

0    with all this stuff going down at the moment w...
1    \the classic war of the worlds\" by timothy hi...
2    the film starts with a manager (nicholas bell)...
3    it must be assumed that those who praised this...
4    superbly trashy and wondrously unpretentious 8...
Name: review_lower, dtype: str

#### Exercise 2.b Remove punctuation and special characters from the text.
This involved created a function with clean_text, and adding a new column "review_clean" to store the updated data. 

In [11]:
import re

#Function to remove punctuation and special characters
def clean_text(text):
    
    #Replace anything not a letter or space with a space
    text = re.sub(r"[^a-zA-Z\s]", " ", text)
    
    return text

#Apply Clean function to lowercase text
df["review_clean"] = df["review_lower"].apply(clean_text)

#Output first few rows
df["review_clean"].head()

0    with all this stuff going down at the moment w...
1     the classic war of the worlds   by timothy hi...
2    the film starts with a manager  nicholas bell ...
3    it must be assumed that those who praised this...
4    superbly trashy and wondrously unpretentious  ...
Name: review_clean, dtype: str

#### Exercise 2.c Remove stop words.
Stop words are common words such as "the", "and", and "is" that do not add meaningful value to the analysis. 
This will need the NLKT package for stopwords to be identified. 
After the stopwords are removed the data will be stored in a new column called "review_no_stopwords".

In [19]:
from nltk.corpus import stopwords

#Load English stopwords
stop_words = set(stopwords.words('english'))

#Function to remove stopwords
def remove_stopwords(text):
    
    #Split text into words
    words = text.split()
    
    #Keep only words not in stopwords list
    filtered_words = [word for word in words if word not in stop_words]
    
    #Bring words back into single string
    return " ".join(filtered_words)

#Apply function to cleaned text
df["review_no_stopwords"] = df["review_clean"].apply(remove_stopwords)

#Output first few rows
df["review_no_stopwords"].head()

0    stuff going moment mj started listening music ...
1    classic war worlds timothy hines entertaining ...
2    film starts manager nicholas bell giving welco...
3    must assumed praised film greatest filmed oper...
4    superbly trashy wondrously unpretentious explo...
Name: review_no_stopwords, dtype: str

#### Exercise 2.d Apply NLTK’s PorterStemmer.
Stemming data reduces word to their root form, which will help in grouping data. 
The Stemmed data will be put in a new column called "review_stemmed".
The output in this column will look odd if you are not familiar with the PorterStemmer. 

In [14]:
from nltk.stem import PorterStemmer

#Create PorterStemmer object
stemmer = PorterStemmer()

#Function to apply stemming to each review
def stem_text(text):
    
    #Split review into individual words
    words = text.split()
    
    #Apply stemming to each word
    stemmed_words = [stemmer.stem(word) for word in words]
    
    #Bring stemmed words back into single string
    return " ".join(stemmed_words)

#Apply stemming to the stopword-removed text
df["review_stemmed"] = df["review_no_stopwords"].apply(stem_text)

#Output the first few rows
df["review_stemmed"].head()

0    stuff go moment mj start listen music watch od...
1    classic war world timothi hine entertain film ...
2    film start manag nichola bell give welcom inve...
3    must assum prais film greatest film opera ever...
4    superbl trashi wondrous unpretenti exploit hoo...
Name: review_stemmed, dtype: str

#### Exercise 2.e * Create a bag-of-words matrix from your stemmed text (output from (4)), where each row is a word-count vector for a single movie review (see sections 5.3 & 6.8 in the Machine Learning with Python Cookbook).

A bag-of-words matrix represents each review as a vector of word counts, 
where each column corresponds to a unique word in the dataset. This will allow the text data to be converted to a numerical format. Each value in the matrix will represent the number of times a word appears in a review.

In [15]:
from sklearn.feature_extraction.text import CountVectorizer

#Initialize CountVectorizer
vectorizer = CountVectorizer()

#Transform stemmed text into bag-of-words matrix
bow_matrix = vectorizer.fit_transform(df["review_stemmed"])

#Output shape of matrix
bow_matrix.shape

(25000, 49642)

#### Exercise 2.f Display the dimensions of your bag-of-words matrix. The number of rows in this matrix should be the same as the number of rows in your original data frame.
For this data, each row represents a movie review and each column represents a unique word in the dataset. The dimensions from the above are 25,000 Rows, 49,642 Columns - Unique words. Therefor the 25,000 rows matches the original dataset. 


#### Exercise 2.g Create a term frequency-inverse document frequency (tf-idf) matrix from your stemmed text, for your movie reviews (see section 6.9 in the Machine Learning with Python Cookbook).

Lastly, creating a TF-IDF (term frequency–inverse document frequency) matrix from the stemmed. Unlike the bag-of-words matrix, TF-IDF reflects the importance of each word by reducing the weight of common words and increasing the importance of less frequent words.



In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer

#Initialize TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()

#Transform stemmed text
tfidf_matrix = tfidf_vectorizer.fit_transform(df["review_stemmed"])

#Output shape of matrix
tfidf_matrix.shape

(25000, 49642)

#### Exercise 2.h Display the dimensions of your tf-idf matrix. These dimensions should be the same as your bag-of-words matrix.

The TF-IDF matrix has the same shape as the bag-of-words matrix, with 25,000 rows and 49,642 columns. Each row representing a movie review, and each column representing a unique word.